### PR TITLE
Ignore partial transactions

### PR DESCRIPTION
**Pull request by jderusse**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #98 
| License       | MIT

### CODE REVIEW

File: src/DataCollector/GuzzleCollector.php [hunk 1]
```diff
@@ -64,6 +64,10 @@ public function collect(Request $request, Response $response, \Exception $except
         /** @var RequestInterface $request */
         foreach ($this->history as $request) {
             $transaction = $this->history[$request];
+            if (!isset($transaction['response'])) {
```
Comment (csarrazi) [inline src/DataCollector/GuzzleCollector.php]: Add a newline before the `if` statement.

Comment (csarrazi) [inline src/DataCollector/GuzzleCollector.php]: Add a newline before the `if` statement.

**Comment by csarrazi** [comment]
Add a new test for the new if statement.

File: src/DataCollector/GuzzleCollector.php [hunk 2]
```diff
@@ -64,6 +64,10 @@ public function collect(Request $request, Response $response, \Exception $except
...
+                continue;
```
Comment (csarrazi) [inline src/DataCollector/GuzzleCollector.php]: This won't actually fix things. You may run into cases where you have no response, but still an error.

**Comment by csarrazi** [comment]
Don't worry, I didn't forget you! I was just really busy this week, with MDBW16 :)

File: src/DataCollector/GuzzleCollector.php [hunk 3]
```diff
@@ -167,3 +167,4 @@ public function getName()
         return 'guzzle';
     }
 }
+
```
Comment (csarrazi) [inline src/DataCollector/GuzzleCollector.php]: Could you revert, please?

File: src/GuzzleHttp/Middleware/HistoryMiddleware.php [hunk 4]
```diff
@@ -58,3 +69,4 @@ function ($reason) use ($request, $options) {
         };
     }
 }
+
```
Comment (csarrazi) [inline src/GuzzleHttp/Middleware/HistoryMiddleware.php]: Remove the additional newline at end.

**Comment by csarrazi** [comment]
As mentioned earlier, could you add a new test for checking whether this works as intended, now?

**Comment by csarrazi** [comment]
I'm not really sure about the implications of the `on_stats` approach. I think generating a unique request ID would be the best approach, even if it means that we would need to inject an additional header.

I don't see any problem with injecting a new header, as in prod, the history middleware should be disabled, and we already use the same approach to inject the cache status (hit, miss) in the response.

**Comment by jderusse** [comment]
I replace the SplStorage by a dedicated storage service. 

This reduce coupling between  GuzzleCollector and HistoryMiddleware on the data structure. But add a BC break by changing the the typeHinting.

If you're OK with this kind of implementation, I'll add tests.

**Comment by csarrazi** [comment]
I would have used a simpler approach, actually. I do not really see the point in having a dedicated class to store the info. We could directly store this in an array instead, and leave the logic in the middleware class. This would actually also most likely fix the BC issue (which was actually the case a few months ago).

Otherwise, the logic itself seems correct!

**Comment by jderusse** [comment]
Because the GuzzleCollector::onStats and the HistoryMiddle do almost the same thing but in two different location... I wanted share the same code (correlationId check and fetch) in this storage service.  And enforce consistency by adding checks...

**Comment by jderusse** [comment]
Simplification done and tests added

File: src/GuzzleHttp/Middleware/HistoryMiddleware.php [hunk 5]
```diff
@@ -23,34 +23,32 @@ class HistoryMiddleware
...
+    public function __construct(\ArrayObject $container)
     {
         $this->container = $container;
     }
 
     public function __invoke(callable $handler)
     {
         return function (RequestInterface $request, array $options) use ($handler) {
+            $correlationId = uniqid();
+            $request = $request->withAddedHeader('csa-guzzle-correlation-id', $correlationId);
```
Comment (csarrazi) [inline src/GuzzleHttp/Middleware/HistoryMiddleware.php]: Move the header name in a class constant, and add the `X-` prefix. This is a non-standard header, and should be prefixed by `X-` (eXtended).

See https://github.com/csarrazi/CsaGuzzleBundle/blob/548e35124e4e854f610bcc501d239c6cbda23316/src/GuzzleHttp/Middleware/CacheMiddleware.php#L26 for more info.

File: tests/GuzzleHttp/Middleware/HistoryMiddlewareTest.php [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+        $this->assertCount(2, $storage);
+        $this->assertArrayHasKey('request', $storage[0]);
+        $this->assertArrayHasKey('response', $storage[0]);
+        $this->assertArrayHasKey('options', $storage[0]);
+        $this->assertArrayHasKey('info', $storage[0]);
+        $this->assertArrayHasKey('error', $storage[0]);
+        $this->assertSame($response, $storage[0]['response']);
+    }
+}
```
Comment (csarrazi) [inline tests/GuzzleHttp/Middleware/HistoryMiddlewareTest.php]: You should add a test with a middleware that adds another header, before and/or after the history middleware, to ensure that both scenarios work.

Likewise, you should ensure that an error returns no response, but an error.

File: src/GuzzleHttp/Middleware.php [hunk 7]
```diff
@@ -35,7 +35,7 @@ public static function cache(StorageAdapterInterface $adapter, $debug = false)
         return new CacheMiddleware($adapter, $debug);
     }
 
-    public static function history(\SplObjectStorage $container)
+    public static function history(\ArrayObject $container)
```
Comment (csarrazi) [inline src/GuzzleHttp/Middleware.php]: We could simply use a reference to an array instead.

File: src/DataCollector/GuzzleCollector.php [hunk 8]
```diff
@@ -41,17 +41,34 @@ class GuzzleCollector extends DataCollector
     public function __construct($maxBodySize = self::MAX_BODY_SIZE)
     {
         $this->maxBodySize = $maxBodySize;
-        $this->history = new \SplObjectStorage();
+        $this->history = new \ArrayObject();
```
Comment (csarrazi) [inline src/DataCollector/GuzzleCollector.php]: Array here, and change the `getHistory()` method to return a reference to the array.

File: tests/GuzzleHttp/MiddlewareTest.php [hunk 9]
```diff
@@ -25,7 +25,7 @@ public function testCacheMiddleware()
 
     public function testHistoryMiddleware()
     {
-        $this->assertInstanceOf(Middleware\HistoryMiddleware::class, Middleware::history(new \SplObjectStorage()));
+        $this->assertInstanceOf(Middleware\HistoryMiddleware::class, Middleware::history(new \ArrayObject()));
```
Comment (csarrazi) [inline tests/GuzzleHttp/MiddlewareTest.php]: Reference to array.

File: src/GuzzleHttp/Middleware.php (hunk 7 above)
Comment (jderusse) [inline src/GuzzleHttp/Middleware.php]: That's a lot of reference. The code wouldn't be simpler but more complex to understand

**Comment by jderusse** [comment]
Improve test and header's constant
.
I wouldn't replace the storage object by an array, given it force to use reference in several method signature (`Middleware::history`, `GuzzleCollector::getHistory`, `HistoryMiddleware::__construct`) which is less readable IMHO, specially in the return signature of getHistory.

**Comment by csarrazi** [comment]
Could you rebase your PR? This should fix some style issues.

**Comment by jderusse** [comment]
rebased and CS fixed. ready to merge ;)

**Pull request merged**